MDM docs: Add end user migration workflow (#13261)
fleetdm/fleet: docs/Using Fleet/MDM-migration-guide.md

1. Prepare to migrate hosts
2. Choose migration workflow and migrate hosts

### Step 1: prepare to migrate hosts

1. Connect Fleet to Apple Business Manager (ABM). Learn how [here](./MDM-setup.md#apple-business-manager-abm).
2. [Enroll](./Adding-hosts.md) your hosts to Fleet with [Fleetd and Fleet Desktop](https://fleetdm.com/docs/using-fleet/adding-hosts#including-fleet-desktop) 
3. Ensure your end users have access to an admin account on their Mac. End users won't be able to migrate on their own if they have a standard account.
4. Migrate your hosts to Fleet in ABM:
    1. In ABM, unassign the existing hosts' MDM server from the old MDM solution: In ABM, select **Devices** and then select **All Devices**. Then, select **Edit** next to **Edit MDM Server**, select **Unassign from the current MDM**, and select **Continue**.
    2. In ABM, assign these hosts' MDM server to Fleet: In ABM, select **Devices** and then select **All Devices**. Then, select **Edit** next to **Edit MDM Server**, select **Assign to the following MDM:**, select your Fleet server in the dropdown, and select **Continue**.

### Step 2: choose migration workflow and migrate hosts

> There are two migration workflows in Fleet: default and end user.

Author: Hmm, yeah I was considering these options:
1. "IT initiated" and "End user initiated"
2. "Default workflow" and "End user workflow"
3. "Default" and "end user" (current)

I liked "IT initiated" and "End user initiated" because the difference between the workflows is more explicit. 

However, we call the end user migration workflow feature "end user migration workflow" in the product. We don't use the work "initiated."

I went with #3 because I think "default" tells the user that this will be the workflow used if they don't do any additional set up. "Default" is shorter than "Default workflow."

What do you think?
I aldo

> The default migration workflow requires that the IT admin unenrolls hosts from the old MDM solution before the end user can complete migration. No additional set up is needed for this workflow.

Author: Makes sense. I updated this section based on your feedback.

The end user migration workflow allows the end user to kick-off migration by unenrolling from the old MDM solution on their own. Once the user is unenrolled, they're nagged to turn on MDM features in Fleet. This reduces the gap in MDM coverage.

#### Default workflow

> 1. In your old MDM solution, unenroll the hosts to be migrated. MacOS does not allow multiple MDMs to be installed at once.

Author: We include that here in the prepare step: https://github.com/fleetdm/fleet/pull/13261/files#diff-83345688042504b8f2ec9fbfdb07bac7d1c90210f4deaf5fff85989c3c8372daR50

    * If you don’t see the yellow banner or the **Turn on MDM** button, select the purple **Refetch** button at the top of the page. 
    * If you still don't see the **Turn on MDM** button or the **My device** page presents you with an error, please contact your IT administrator.

<img width="1400" alt="My device page - turn on MDM" src="https://user-images.githubusercontent.com/5359586/229950406-98343bf7-9653-4117-a8f5-c03359ba0d86.png">

#### End user workflow

> Available in Fleet Premium or Ultimate

The end user migration workflow is supported for automatically enrolled (DEP) hosts.

To watch a GIF that walks through the end user experience for the end user migration workflow, in the Fleet UI, head to **Settings > Integrations > Mobile device management (MDM)**, and scroll down to the **End user migration workflow** section.

In Fleet, you can configure the end user workflow using the Fleet UI or fleetctl command-line tool.

Fleet UI:

1. Select the avatar on the right side of the top navigation and select **Settings > Integrations > Mobile device management (MDM)**.

2. Scroll down to the **End user migration workflow** section and select the toggle to enable the workflow.

> 3. Under **Mode** choose a mode and enter the webhook URL for you automation tool (ex. Tines) under **Webhook URL** and select **Save**.

Author: I'm leaning on the UI to explain this to the user (they'll be in the UI as they follow along):

![Screenshot 2023-08-10 at 3 22 37 PM](https://github.com/fleetdm/fleet/assets/47070608/9dcfbc17-6ac1-42f4-a6d5-7ed12597fa6a)

    1. Head to **Settings** > **Organization settings** >
**Organization info** and add URLs to your logos in the **Organization avatar URL(for dark backgrounds)** and **Organization avatar URL (for light backgrounds)** fields.
    
    2. Select **Save**.

fleetctl CLI:

1. Create `fleet-config.yaml` file or add to your existing `config` YAML file:

```yaml
apiVersion: v1
kind: config
spec:
  mdm:
    macos_migration:
      enable: true
      mode: "voluntary"
      webhook_url: "https://example.com"
  ...
```

2. Fill in the above keys under the `mdm.macos_migration` key. 

To learn about each option, in the Fleet UI, select the avatar on the right side of the top navigation, select **Settings > Integrations > Mobile device management (MDM)**, and scroll down to the **End user migration workflow** section.

3. During the end user migration workflow, an end user's device will have their selected system theme (light or dark) applied. If your logo does not look good on either light or dark backgrounds, you can optionally set a logo for each theme:

```yaml
apiVersion: v1
kind: config
spec:
  org_info:

>     org_logo_url: https://fleetdm.com/images/press-kit/fleet-blue-logo.png

Reviewer: Wait, so this is the default logo AND it's the black theme logo? And only if you set ligght_background logo below, then you get the light logo? Is that right? It does not seem ocnsistent with the text in bullet 3.

Reviewer: Another way to ask this is, why isn't there three logo options?
1. org_logo_url
2. org_logo_light_background_url
3. org_logo_dark_background_url

Author: > default logo AND it's the black theme logo?

Not quite. The Fleet logo is the default for `org_logo_url` and `org_logo_light_background_url `

If the user updates `org_logo_url` this only updates the logo in the Fleet top navigation and black theme:

![Screenshot 2023-08-10 at 3 26 03 PM](https://github.com/fleetdm/fleet/assets/47070608/7e8de57b-02a6-4c1d-8901-31891b9ba2dd)

If the user updates `org_logo_light_background_url `, this only updates the logo in the light theme. The Fleet logo is still used in top nav and dark theme.

I updated the docs to clarify this.

Reviewer: I think it needs sentence here that's like, "if you only set `org_logo_url`, then it will be used for both dark and light mode"

Author: This isn't the behavior. Good feedback tho. I updated the docs to further clarify:

"Add URLs to your logos that look good on a dark background and light background in the `org_logo_url` and `org_logo_url_light_background` keys respectively. If you only set a logo for one, the Fleet logo will be used for the other."

    org_logo_url_light_background: https://fleetdm.com/images/press-kit/fleet-white-logo.png
  ...
```

Add URLs to your logo that looks good on dark background and light backgrounds in the `org_logo_url` and `org_logo_url_light_background` keys respectively.

4. Run the fleetctl `apply -f fleet-config.yml` command to add your configuration.

5. Confirm that your configuration was saved by running `fleetctl get config`.

6. Send [these guided instructions](#how-to-turn-on-mdm-end-user) to your end users to complete the final few steps via Fleet Desktop.

##### How to turn on MDM (end user)

1. Select the Fleet icon in your menu bar and select **Migrate to Fleet**.

2. Select **Start** in the **Migrate to Fleet** popup. 

2. On your **My device** page, select **Turn on MDM** the button in the yellow banner and follow the instructions. 
    * If you don’t see the yellow banner or the **Turn on MDM** button, select the purple **Refetch** button at the top of the page. 
    * If you still don't see the **Turn on MDM** button or the **My device** page presents you with an error, please contact your IT administrator.

> ## Migration report

Reviewer: This section is great. It answers the "did it work?" question. I wonder if we can be more explicit here. "Migration report" is so bland and abstract. "How do I check the progress of my migration?" "Checking migration progress" etc etc. Let's use clear everyday language as much as possible.

Author: Agree it's bland. Updated the docs to use "Check migration progress:

1. A [deployed Fleet instance](../Deploying/Introduction.md)
2. [Fleet connected to Apple](./MDM-setup.md)

## Migrate manually enrolled hosts

1. [Enroll](./Adding-hosts.md) your hosts to Fleet with [Fleetd and Fleet Desktop](https://fleetdm.com/docs/using-fleet/adding-hosts#including-fleet-desktop) 
2. Ensure your end users have access to an admin account on their Mac. End users won't be able to migrate on their own if they have a standard account.
3. In your old MDM solution, unenroll the hosts to be migrated. MacOS does not allow multiple MDMs to be installed at once.
4. Send [these guided instructions](#how-to-turn-on-mdm) to your end users to complete the final few steps via Fleet Desktop.
    * Note that there will be a gap in MDM coverage between when the host is unenrolled from the old MDM and when the host turns on MDM in Fleet.

### End user experience

1. On their **My device** page, once an end user's device is unenrolled from the old MDM solution, the end user will be given the option to manually download the MDM enrollment profile.
   

> 2. Once downloaded, the user will receive a system notification that the Device Enrollment profile has been needs to be installed in the System Settings &gt; Profiles section. 

Author: Great catch. Thanks

   
3. After installation, the MDM enrollment profile can be removed by the end user at any time.

### How to turn on MDM

1. Select the Fleet icon in your menu bar and select **My device**.

![Fleet icon in menu bar](https://raw.githubusercontent.com/fleetdm/fleet/main/website/assets/images/articles/fleet-desktop-says-hello-world-cover-1600x900@2x.jpg)

2. On your **My device** page, select **Turn on MDM** the button in the yellow banner and follow the instructions. 
  - If you don’t see the yellow banner or the **Turn on MDM** button, select the purple **Refetch** button at the top of the page. 
  - If you still don't see the **Turn on MDM** button or the **My device** page presents you with an error, please contact your IT administrator.

<img width="1400" alt="My device page - turn on MDM" src="https://user-images.githubusercontent.com/5359586/229950406-98343bf7-9653-4117-a8f5-c03359ba0d86.png">

## Migrate automatically enrolled (DEP) hosts

> Automatic enrollment is available in Fleet Premium or Ultimate

To migrate automatically enrolled hosts, we will do the following steps:

1. Prepare to migrate hosts

> 2. Choose migration workflow and migrate hosts

Author: I'm going to leave it in because I think it gives the user a heads up that they'll have to choose a migration workflow later. So, if they catch that, they can look ahead at the options.

3. Ensure your end users have access to an admin account on their Mac. End users won't be able to migrate on their own if they have a standard account.
4. Migrate your hosts to Fleet in ABM:
    1. In ABM, unassign the existing hosts' MDM server from the old MDM solution: In ABM, select **Devices** and then select **All Devices**. Then, select **Edit** next to **Edit MDM Server**, select **Unassign from the current MDM**, and select **Continue**.
    2. In ABM, assign these hosts' MDM server to Fleet: In ABM, select **Devices** and then select **All Devices**. Then, select **Edit** next to **Edit MDM Server**, select **Assign to the following MDM:**, select your Fleet server in the dropdown, and select **Continue**.

### Step 2: choose migration workflow and migrate hosts

There are two migration workflows in Fleet: default and end user.

The default migration workflow requires that the IT admin unenrolls hosts from the old MDM solution before the end user can complete migration. This will result in a gap in MDM coverage until the end user completes migration.

> The end user migration workflow allows the end user to kick-off migration by unenrolling from the old MDM solution on their own. Once the user is unenrolled, they're nagged to turn on MDM features in Fleet. This reduces the gap in MDM coverage.

Author: I like "prompted" (I changed it).

Configuring the end user migration workflow requires a few additional steps.

#### Default workflow

1. In your old MDM solution, unenroll the hosts to be migrated. MacOS does not allow multiple MDMs to be installed at once.

2. Send [these guided instructions](#how-to-turn-on-mdm-default) to your end users to complete the final few steps via Fleet Desktop.
    * Note that there will be a gap in MDM coverage between when the host is unenrolled from the old MDM and when the host turns on MDM in Fleet.

##### End user experience

1. The end user will receive a "Device Enrollment: &lt;organization&gt; can automatically configure your Mac." system notification within the macOS Notifications Center. 
   
2. After the end user clicks on the system notification, macOS will open the "Profiles" System Setting and ask the user to "Allow Device Enrollment: &lt;organization&gt; can automatically configure your Mac based on settings provided by your System Administrator."
  

> 3. If the end user does not install the profile, the system notification will continue to nag the end user until the setting has been allowed.

Author: Used "prompt"

   
4. Once this setting has been approved, the MDM enrollment profile cannot be removed by the end user.

##### How to turn on MDM (default)

1. Select the Fleet icon in your menu bar and select **My device**.

![Fleet icon in menu bar](https://raw.githubusercontent.com/fleetdm/fleet/main/website/assets/images/articles/fleet-desktop-says-hello-world-cover-1600x900@2x.jpg)

> 2. On your **My device** page, select **Turn on MDM** the button in the yellow banner and follow the instructions. 

Author: ✅

    * If you don’t see the yellow banner or the **Turn on MDM** button, select the purple **Refetch** button at the top of the page. 
    * If you still don't see the **Turn on MDM** button or the **My device** page presents you with an error, please contact your IT administrator.

<img width="1400" alt="My device page - turn on MDM" src="https://user-images.githubusercontent.com/5359586/229950406-98343bf7-9653-4117-a8f5-c03359ba0d86.png">

#### End user workflow

> Available in Fleet Premium or Ultimate

The end user migration workflow is supported for automatically enrolled (DEP) hosts.

> To watch a GIF that walks through the end user experience for the end user migration workflow, in the Fleet UI, head to **Settings > Integrations > Mobile device management (MDM)**, and scroll down to the **End user migration workflow** section.

Author: Great catch. I removed the repetition.

In Fleet, you can configure the end user workflow using the Fleet UI or fleetctl command-line tool.

Fleet UI:

1. Select the avatar on the right side of the top navigation and select **Settings > Integrations > Mobile device management (MDM)**.

2. Scroll down to the **End user migration workflow** section and select the toggle to enable the workflow.

3. Under **Mode** choose a mode and enter the webhook URL for you automation tool (ex. Tines) under **Webhook URL** and select **Save**.

> 4. During the end user migration workflow, an end user's device will have their selected system theme (light or dark) applied. If your logo does not look good on either light or dark backgrounds, you can optionally set a logo for each theme:

Reviewer 2: "both light and dark" might be clearer. Also, may consider an alternative to "doesn't look good" just to avoid negative-sounding language about the user's logo:
```suggestion
4. During the end user migration workflow, an end user's device will have their selected system theme (light or dark) applied. If your logo is not easy to see on both light and dark backgrounds, you can optionally set a logo for each theme:
```

    1. Head to **Settings** > **Organization settings** >
**Organization info** and add URLs to your logos in the **Organization avatar URL(for dark backgrounds)** and **Organization avatar URL (for light backgrounds)** fields.
    
    2. Select **Save**.

fleetctl CLI:

1. Create `fleet-config.yaml` file or add to your existing `config` YAML file:

```yaml
apiVersion: v1
kind: config
spec:
  mdm:
    macos_migration:
      enable: true

>       mode: "voluntary"

Reviewer 2: I think we should omit quotes if there aren't special characters that need escaping, since I don't see us using them in other examples

Author: Hmm, I think we need quotes for "voluntary" in this example because `mode` expects a string. @gillespi314 is that right?

Reviewer 2: My understanding is that YAML strings only need quotes when escaping certain special characters. Here's the team config as an example, many strings aren't in quotes:
![Screenshot 2023-08-15 at 6 09 42 PM](https://github.com/fleetdm/fleet/assets/3065949/9ae456fd-30cd-4c81-b164-abe9f8abb48a)

That said, don't want this to block merging since that syntax isn't _incorrect_— just thinking about conventions for these examples going forward.

      webhook_url: "https://example.com"
  ...
```

2. Fill in the above keys under the `mdm.macos_migration` key. 

To learn about each option, in the Fleet UI, select the avatar on the right side of the top navigation, select **Settings > Integrations > Mobile device management (MDM)**, and scroll down to the **End user migration workflow** section.

3. During the end user migration workflow, the window will show the Fleet logo on top of a dark and light background (appearance configured by end user).

If want to add a your organization's logo, you can optionally set a logo for each background:

```yaml
apiVersion: v1
kind: config
spec:
  org_info:
    org_logo_url: https://fleetdm.com/images/press-kit/fleet-blue-logo.png
    org_logo_url_light_background: https://fleetdm.com/images/press-kit/fleet-white-logo.png
  ...
```

> Add URLs to your logos that look good on a dark background and light background in the `org_logo_url` and `org_logo_url_light_background` keys respectively. If you only set a logo for one, the Fleet logo will be used for the other.

Reviewer 2: See previous comment re: "look good on":
```suggestion
Add URLs to your logos that are visible on a dark background and light background in the `org_logo_url` and `org_logo_url_light_background` keys respectively. If you only set a logo for one, the Fleet logo will be used for the other.
```

4. Run the fleetctl `apply -f fleet-config.yml` command to add your configuration.

5. Confirm that your configuration was saved by running `fleetctl get config`.

6. Send [these guided instructions](#how-to-turn-on-mdm-end-user) to your end users to complete the final few steps via Fleet Desktop.

##### How to turn on MDM (end user)

1. Select the Fleet icon in your menu bar and select **Migrate to Fleet**.

2. Select **Start** in the **Migrate to Fleet** popup. 

> 2. On your **My device** page, select **Turn on MDM** the button in the yellow banner and follow the instructions. 

Reviewer 2: ```suggestion
2. On your **My device** page, select the **Turn on MDM** button in the yellow banner and follow the instructions. 
```
